### DISCUSSION TITLE
plot_question

### DISCUSSION THREAD

**Discussion by rtwod2** [pyg-team/pytorch_geometric#4740] [Q&A]
Hi all,

I'm training a GNN model using pytorch geometric for a regression problem. I'm using cross-validation. The mean plot for my target is the below one. Do you find it okay?

![mean_trainval_loss_exectime_gpu_20220530_160346](https://user-images.githubusercontent.com/98180692/171006058-6e4d09b9-514f-474b-b413-6c8c887d08fa.png)

Thank you in advance for your time and consideration of my question.

**Comment by rusty1s** [comment]
This is hard to tell just by looking at the loss. I suggest you make use of some additional metrics to judge the quality of your model, e.g., accuracy, F1-score, etc, and compare it to baselines as well.

**Reply by rtwod2** [reply]
Thank you for answering. I was thinking about using other metrics. 
But this is a regression problem and I'm not aware if for example F1-score can be applied in regression. Could you please point me to something similar using i.e. F1-score in regression?

Thank you so much for your help.

**Reply by rusty1s** [reply]
For regression, you should probably want to report MAE and MSE.

**Reply by rtwod2** [reply]
Yes, exactly. So, I'm just using mae and mse. 
I'm attaching you some results from the model:

the predictions in the test set are:                     
tensor(0.3083)  
tensor(0.3083)  
tensor(0.3083) 
tensor(0.3083) 
tensor(0.3083)  
tensor(0.3083)  
tensor(0.3083)  
tensor(0.3083) 
tensor(0.3083)
tensor(0.3083) 
tensor(0.3083)  
tensor(0.3083)  
tensor(0.3083)  
tensor(0.3083)  

the labels in the test set are:
tensor(0.2319) 
tensor(0.2183) 
tensor(0.2491)  
tensor(0.2247)  
tensor(0.2325)
tensor(0.2794)  
tensor(0.2640)  
tensor(0.2434)  
tensor(0.2607)  
tensor(0.2423)  
tensor(0.2896)  
tensor(0.2688)  
tensor(0.2650) 
tensor(0.2624)

and the train and val loss plot of this model is the below:
![gpu_mloss_exectime_nofstatic_20220531_203228_crossval](https://user-images.githubusercontent.com/98180692/171415141-63fc9961-5321-48c3-b1b8-f89ebd981d75.png)

Do you find these results okay? Cause I'm not sure of the stuck values on predictions (0.3083). 

P.S. I had a standardization of these values , using this formula:
                       scaled = (np.asarray(data) - meand) / std
       and to make the inversed values after training I used the same std, meand:
                       original_values_b = (y_scaled * std) + meand

**Reply by rusty1s** [reply]
Does that mean that the model predicts the same value for each graph. I assume this is definitely undesired.
Your standardization technique and inverse formulation looks correct to me :)

**Reply by rtwod2** [reply]
Yes, this is my thinking too. But, the peculiarity here is that for example the above results are of the same kernel (because my dataset is kernels-ir code). So, for every kernel I have multiple runs, which differ from each other on only one feature.

To make it more clearer to you, In my dataset I have a kernel, which can be a pytorch geometric graph, and I can have many graphs out of one kernel which are going to have node features (type, block, inputbytes). These features are the same for the graphs corresponded to this kernel, except of inputbytes. Is this possible that this is the reason that the model predicts the same value for each graph, from the same kernel? 

Thank you again. Without exaggeration, if you exclude random internet search, you are my only mentor for many months.

**Reply by rusty1s** [reply]
Yes, this sounds like a problem. Ideally, you would combine features and learn from them together. There is not much reasoning a GNN can do when only applied on a single feature (and features are potentially even shared among nodes). Let me know if I misunderstand something :)

**Reply by rtwod2** [reply]
Yes I understand what you’re saying. So to validate what you’re saying and to be more clear for me, holistically.

I’m having x graphs, which correspond to y kernels, and for each kernel I have more than one graph. The graphs correspond for example, to y_i kernel have the same features (type, block) and one different feature (input byte). 

For the above situation of my dataset , you’re saying , that having some features  the same (for ‘groups’ of graphs) , is a problem for the GNN, because it’s eventually tries to predict based on a single feature, the input bytes (the different one).

What could be done for solving this problem?

Thank you so much.

**Reply by rusty1s** [reply]
I think I misunderstood. It looks like you already have a set of features, while for different kernels only a single feature changes (I thought you only have one feature for a single graph). Correct? In this case, this looks fine to me, and the problem of bad model performance is likely stemming from something else.

**Reply by rtwod2** [reply]
Yes, I have already a set of features, but the same kernels have different graphs, with the only different feature to be the input bytes (all the other features are the same for the same kernels). While different kernels have different all the features in between. Is this more clear?

So, I 'll give you an example of some predictions (depicts one of my targets - the plot above is about the other target):
**labels**
16  tensor(15802.5859)  rodinia-3.1-streamcluster-memset_kernel.ll
17  tensor(16260.0508)  rodinia-3.1-streamcluster-memset_kernel.ll
18  tensor(16717.5176)  rodinia-3.1-streamcluster-memset_kernel.ll
19   tensor(9398.0723)                   shoc-1.1.5-Triad-Triad.ll
20  tensor(11227.9336)                   shoc-1.1.5-Triad-Triad.ll
21  tensor(13057.7930)                   shoc-1.1.5-Triad-Triad.ll

**predictions**
16  tensor(25941.0840)  rodinia-3.1-streamcluster-memset_kernel.ll
17  tensor(25941.0840)  rodinia-3.1-streamcluster-memset_kernel.ll
18  tensor(25941.0840)  rodinia-3.1-streamcluster-memset_kernel.ll
19  tensor(26215.4453)                   shoc-1.1.5-Triad-Triad.ll
20  tensor(26215.4453)                   shoc-1.1.5-Triad-Triad.ll
21  tensor(26215.4453)                   shoc-1.1.5-Triad-Triad.ll

For the above, I have the same prediction-number for all the same kernels, (which differ on only one feature) and different prediction-number in between the different kernels.

I have tried to change my model, I used relu, then elu, I changed my dropout rate, I used topkpooling and then I erased topkpooling. But I can see the same behavior on all these experiments.

Do you have any intuition on what's going on?

**Reply by rusty1s** [reply]
Can you show me the model that you are using?

**Reply by rtwod2** [reply]
Yes, of course! Sorry for the late answer, I just saw your feedback. 

``` python3
class GTModel(nn.Module):
    def __init__(self, num_layers, hidden_dim, heads, feat_dropout, top_k_pool, norm, beta=True):

        super().__init__()

        dense_neurons = 256
        self.top_k_every_n = top_k_pool # self.top_k_every_n
        self.num_layers = num_layers
        self.feat_dropout = feat_dropout
        self.norm = None

        self.conv_layers = nn.ModuleList([])
        self.transf_layers = nn.ModuleList([])
        self.pooling_layers = nn.ModuleList([])
        self.norm_layers = nn.ModuleList([])

    
        #we don't calculate the embeddings here, but we calculate them using the NodeFeatureEncoder and then pass them into our  
        #GNN model directly.
        #This is happening per batch of graphs!


        # Transformation Layer
        self.conv1 = TransformerConv(hidden_dim, hidden_dim, heads, dropout=0.3, edge_dim=hidden_dim, norm=self.norm, beta=False)
        self.bn1 = nn.BatchNorm1d(hidden_dim*heads)
        self.transf1 = Linear(hidden_dim*heads, hidden_dim)

         for i in range(self.num_layers):
            self.conv_layers.append(TransformerConv(hidden_dim, hidden_dim, heads, dropout=0.3, edge_dim=hidden_dim, norm=self.norm, beta=False))
            self.norm_layers.append(nn.BatchNorm1d(hidden_dim*heads))
            self.transf_layers.append(Linear(hidden_dim*heads, hidden_dim))         
            if i % self.top_k_every_n == 0:
                self.pooling_layers.append(TopKPooling(hidden_dim, ratio=0.5))    

         self.linear1 = Linear(hidden_dim*2, dense_neurons)
         self.linear2 = Linear(dense_neurons, int(dense_neurons/2))  
         self.linear3 = Linear(int(dense_neurons/2), 1) 

 
    def forward(self, batch_graphs): 
        out = []

        nodefeature = NodeFeatureEncoder.FeatureEncoder(batch_graphs, emb_dim=64) # emb_dim=hidden_dim=64
        renewed_data = nodefeature(batch_graphs) # forward pass

        edgefeature = EdgeFeatureEncoder.eFeatureEncoder(renewed_data, emb_dim=64)
        renewed_data = edgefeature(renewed_data) 

        x, edge_attr, edge_index, batch_index= renewed_data.x, renewed_data.edge_attr, renewed_data.edge_index, renewed_data.batch

        # Initial Transformation
        x = self.conv1(x, edge_index, edge_attr)
        x = self.bn1(x)
        x = torch.nn.functional.elu(self.transf1(x))
        
        global_representation = []

        for i in range(self.num_layers):
            x = self.conv_layers[i](x, edge_index, edge_attr)
            x = self.norm_layers[i](x)
            x = torch.nn.functional.elu(self.transf_layers[i](x))            
            if i % self.top_k_every_n == 0 or i == self.num_layers:
                x, edge_index, edge_attr, batch_index, _, _ = self.pooling_layers[int(i/self.top_k_every_n)](x, edge_index, edge_attr, batch_index) 
            
                global_representation.append(torch.cat([gmp(x, batch_index), gap(x, batch_index)], dim=1))

        x = sum(global_representation)

        print(f'out shape after gap, cat and sum {x.shape}') #[batchsize, hiddendim*lendatalist]

        # Output block
        x = torch.nn.functional.elu(self.linear1(x))
        x = F.dropout(x, p=0.8, training=self.training)
        x = torch.nn.functional.elu(self.linear2(x))
        x = F.dropout(x, p=0.8, training=self.training)
        x = self.linear3(x)

       
        return x
```

And for TransformerConv, I'm using the one available from Pytorch Geometric.

**Reply by rusty1s** [reply]
Thanks. I think your model looks good but also is pretty complex. Maybe we should start with a more simpler baseline, e.g., a simple `MLP` or a simple `GraphSAGE` model with `global_mean_pooling` at the end?

**Comment by rtwod2** [comment]
Thank you so much for everything. Sorry for my late reply, but I've had a lot going on. I'm looking at the GraphSage paper, and I'll implement this model to see if it can be improved. I'm thinking if it might help to have position encodings. I'll look into them, and send you anything new that comes up. 
Thank you soooo much.
